### ISSUE TITLE
Support ties across measures with (changing) polyphony

### ISSUE THREAD

**Issue by th-we** [th-we/MEILER#7]
@uliska wrote:
> Das könnte ein nicht ganz triviales Problem werden. Grundsätzlich können
Spanner ausschließlich innerhalb eines Voice-Kontexts gesetzt werden.
Bei temporärer Polyphonie (<< >>) gibt es die Möglichkeit, alle Stimmen
temporär neu zu erzeugen (\\) oder die erste Stimme durchgehen zu lassen
(\new Voice). Nur mit der letzteren Variante kann man in *einer* Stimme
Bögen etc. setzen. Sofern Bögen in mehr als einer Stimme nötig sind,
*muss* man den polyphonen Abschnitt erweitern, so dass die Stimmen
jeweils kontinuierlich sind.
> 
> Das Problem ist, dass zumindest im MEI, wahrscheinlich aber schon in
Sibelius, die Takte als Boxen verstanden werden. Genauso überträgt
MEILER das ja auch, mit einem << >>-Ausdruck pro Takt.
> 
> Wenn man das nicht grundsätzlich ändern - also eine
"Layer"/"Voice"-basierte Stimme erzeugen - kann, muss zumindest
gewährleistet werden, dass zu haltende Noten immer in der gleichen, und
zwar ersten, Stimme zu stehen kommen. Ich weiß nicht, ob das durchgehend
möglich ist.
>
> Der Workaround ist im allgemeinen, eine versteckte Stimme nur für die
Bögen zu machen. Das ist aber relativ komplex und meiner Erfahrung nach
nicht zu automatisieren.

**Comment by th-we** [comment]
@uliska Could you elaborate more on how to define ties in invisible voices?

**Comment by uliska** [comment]
OK, as we shouldn't discriminate on preexisting conditions ;-) I'll stick to the concept of having each measure boxed in a `<< >>`. The orignal situation is this:

```lilypond
\new Voice {
  <<
    { 
      \voiceOne
      c''2 c''2 ~
    }
    \\
    {
      \voiceTwo
      c'1
    }
  >>
  <<
    { 
      \voiceTwo
      c'1
    }
    \\
    {
      \voiceOne
      c''2 c''2 ~
    }
  >>
}
```
 which leads to an `unterminated tie` warning.

![document](https://cloud.githubusercontent.com/assets/1812148/23743712/069c6206-04b2-11e7-886a-f167299da142.png)

The proper way to fix this code is to use the alternative syntax for temporary polyphony (`\new Voice` instead of `\\` and ensure the tied notes are in the first voice:

```lilypond
\new Voice {
  <<
    { 
      \voiceOne
      c''2 c''2 ~
    }
    \new Voice
    {
      \voiceTwo
      c'1
    }
  >>
  <<
    { 
      \voiceOne
      c''2 c''2 ~
    }
    \new Voice
    {
      \voiceTwo
      c'1
    }
  >>
}
```

If that's not possible because either the input can't be organized that way or because there have to be ties in more than one voice an alternative is to use a hidden voice just for the tie(s). This also requires the other syntax and to define that hidden voice in an additional *first* voice.

```lilypond
\new Voice {
  <<
    % an additional voice for printing the tie
    {
      % set up like the "real" voice
      \voiceOne
      s2 
      % hiding the notes is not always necessary but usually a good thing
      \once \hideNotes
      % coloring the tie is only for demonstration to show 
      % that it is *this* one that gets printed
      \once \override Tie.color = #red
      c''2 ~
    }
    % \new Voice command makes the previous voice continue before and after
    % and add the following voice as temporary.
    \new Voice
    { 
      \voiceOne
      % note that the following tie is *not* printed because it 
      % doesn't have a complement in the same voice.
      % Interestingly LilyPond doesn't issue a warning anymore.
      c''2 c''2 ~
    }
    \new Voice
    {
      \voiceTwo
      c'1
    }
  >>
  <<
    % extra first voice to complete the tie.
    {
      \voiceOne
      \once \hideNotes
      c''2 s2
    }
    \new Voice
    { 
      \voiceTwo
      c'1
    }
    \new Voice
    {
      \voiceOne
      c''2 c''2 ~
    }
  >>
}
```
![document](https://cloud.githubusercontent.com/assets/1812148/23744295/f71f5886-04b3-11e7-91d0-32d8bf7d51dd.png)

**Comment by th-we** [comment]
@uliska As long as the order and number of voices does not change, ties seem to work as expected - or is there anything I am missing?

    {
      << 
        { 
          a'1~ 
        } 
        \\ 
        { 
          r2 f'2~ 
        } 
      >>
      << 
        { 
          a'1
        }
        \\ 
        { 
          f'2 r2
        }
      >>
    }

![bildschirmfoto vom 2017-03-10 14-24-16](https://cloud.githubusercontent.com/assets/1147152/23796840/53df587a-059d-11e7-9df2-dcf04ecc0ea6.png)

By automatically adding empty dummy voices where necessary, this could get us quite far without the need for manual intervention.

**Comment by uliska** [comment]
I'm surprised by this example but it seems correct. Either *I* have misunderstood something or the language generally used to describe this situation is ambigous´or there have been changes recently that I missed.

I'll investigate as far as the project is concerned (and maybe generally).

**Comment by uliska** [comment]
OK, I've checked on my work computer, and it seems right - but it seems to contradict the explicit language in the manual. This states that with the `\\` construct all voices are created in new temporary Voice contexts that are then discarded.

You can check the (documented) behaviour by starting the sample with

```lilypond
{
  a'1 ~
  << 
    { 
      a'1~ 
```

where you'll see that the first tie from the monophonic into the polyphonic section does *not* work. The same is true when you put some monophonic music between the two polyphonic blocks.

It seems like the temporary voices are *not* discarded when they are immediately followed by the same amount of voices. (I'll suggest an update to the manuals).

For our project this means that boxing each measure in ``<< >>`` doesn't necessarily have a penalty as consecutive voices are implicitly continued.

One caveat is that when for example ties span more than one full measure "box" this continuity has to be guaranteed. It seems best if a maximum number of voices is determined and dummy boxes with spacer rests are created throughout the piece. (without guarantee).

**Comment by th-we** [comment]
I'll follow the dummy-voices-with-spacers approach. Because this does not cover all situations, I'll try to log all situations where a tie continues in a different MEI layer.

**Comment by th-we** [comment]
Fixed in b5d433c
